### ISSUE TITLE
Test-drive atomics in balances

### ISSUE THREAD

**Issue by garious** [solana-labs/solana#126]
Here's the hard part:

```rust
    /// Deduct tokens from the 'from' address the account has sufficient
    /// funds and isn't a duplicate.
    pub fn process_verified_transaction_debits(&self, tr: &Transaction) -> Result<()> {
        let bals = self.balances.read().unwrap();

        // Hold a write lock before the condition check, so that a debit can't occur
        // between checking the balance and the withdraw.
        let option = bals.get(&tr.from);
        if option.is_none() {
            return Err(AccountingError::AccountNotFound);
        }
        let mut bal = option.unwrap().write().unwrap();

        if *bal < tr.data.tokens {
            return Err(AccountingError::InsufficientFunds);
        }

        if !self.reserve_signature_with_last_id(&tr.sig, &tr.data.last_id) {
            return Err(AccountingError::InvalidTransferSignature);
        }

        *bal -= tr.data.tokens;

        Ok(())
    }
```

We currently hold a lock on the balance while we check the funds and reserve the signature. If we switched to lock-free, we'd need a way to remove that signature if a `compare_swap()` returned a balance that was too insufficient.

**Comment by garious** [comment]
Per fd5ef94b5a8a944827890a08c26cae8dc1635d5a, this effort should go smoother now. We don't have a demo that measures success very well, so will need a new microbenchmark for this one.

**Comment by garious** [comment]
Fixed in 47c23178102f1083d2c45b2f5b0a9d3c1917da00
